### ISSUE TITLE
ValueError when opening NetCDF file with time:units = "Month since xxxx-xx-xx"

### ISSUE THREAD

**Issue by admg26** [pygeode/pygeode#138]
Following commit c5a68ca5c0ddc7e6e9b41041fbc58e1eb7d1398d, units are now read from the cf metadata of the NetCDF file. This now causes a ValueError when opening netcdf files with monthly mean data. When constructing the `StandardTime` axis, units=months is not a valid option. The code fails when checking whether units has a unitfactor.

Note that the [cf metadata conventions](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#time-coordinate) advise caution when using month as a unit. Should `.open` do something clever with months?
 
```
~/.local/lib/python3.6/site-packages/pygeode-1.4.0-py3.6-linux-x86_64.egg/pygeode/formats/__init__.py in open(filename, format, value_override, dimtypes, namemap, varlist, cfmeta, **kwargs)
    160 
    161   return format.open(filename, value_override=value_override, dimtypes=dimtypes, \
--> 162               namemap=namemap, varlist=varlist, cfmeta=cfmeta, **kwargs)
    163 # }}}
    164 

~/.local/lib/python3.6/site-packages/pygeode-1.4.0-py3.6-linux-x86_64.egg/pygeode/formats/netcdf4.py in open(filename, value_override, dimtypes, namemap, varlist, cfmeta)
    255       # with the same name as the dimension)
    256       dataset = dims2axes(dataset)
--> 257       return finalize_open(dataset, dimtypes, namemap, varlist, cfmeta)
    258   except IOError:  # Problem accessing the file?
    259     raise

~/.local/lib/python3.6/site-packages/pygeode-1.4.0-py3.6-linux-x86_64.egg/pygeode/formats/__init__.py in finalize_open(dataset, dimtypes, namemap, varlist, cfmeta)
    264     # Skip anything that we're going to override in dimtypes
    265     # (so we don't get any meaningless warnings or other crap from cfmeta)
--> 266     dataset = cf.decode_cf(dataset, ignore=list(dimtypes.keys()))
    267 
    268   # Apply custom axis types?

~/.local/lib/python3.6/site-packages/pygeode-1.4.0-py3.6-linux-x86_64.egg/pygeode/formats/cfmeta.py in decode_cf(***failed resolving arguments***)
    410       # Create the time axis
    411       startdate={'year':year, 'month':month, 'day':day, 'hour':hour, 'minute':minute, 'second':second}
--> 412       axisdict[name] = cls(a.values, startdate=startdate, units=res, name=name, atts=atts)
    413       # Special case: start year=0 implies a climatology
    414       #NOTE: 'climatology' attribute not used, since we don't currently keep

~/.local/lib/python3.6/site-packages/pygeode-1.4.0-py3.6-linux-x86_64.egg/pygeode/timeaxis.py in __init__(self, values, datefmt, units, startdate, **kwargs)
    416 
    417     if units not in self.unitfactor.keys():
--> 418       raise ValueError('units ("%s") must be one of the following: %s' % (units, list(self.unitfactor.keys())))
    419 
    420     return Time.__init__(self, values=values, units=units, startdate=startdate, **kwargs)

ValueError: units ("months") must be one of the following: ['seconds', 'minutes', 'hours', 'days']
```

**Comment by neishm** [comment]
Does this not crash before commit c5a68ca was made?  I don't see the `units` variable being passed into the time axis constructor (line 412 of cfmeta.py above), so not sure what difference it would be making in the behaviour.

We do have existing support for a monthly time axis (such as those generated from the `monthlymean` operator), so there could be an argument that pygeode should support reading those files in from disk.  Maybe on the condition that the month intervals are whole numbers?

**Comment by aph42** [comment]
The problem is actually in a new check that I added to the CalendarTime constructor on whether ``units`` is in the list of options for the ``unitfactor`` dictionary. This list includes only options that have a fixed scaling relative to seconds. Before the update, the pygeode would read in the file and create a time axis, but it would (I believe) have broken if you tried plotting or doing anything that needed to process the dates explicitly. 

The cf standard does include the possibility of using 'month' as a unit in calendars, but it's ambiguous how to interpret it. Apparently udunits defines one month to be exactly one twelfth of a year, and one year is the actual orbital period of the earth (~365.25ish days). There is also a common_year and leap_year unit. This is what we'd need to include in a ``unitfactor`` calculation, but it's probably not what is intended in this particular data file. 

It also wouldn't match what we do for monthly means, which is to take a timeaxis and uniquefy it based on the 'month' attribute array; this is much closer to what is probably intended in the SWOOSH dataset, but is a bit murky as to what the actual values of the calendar will be. Computing ``v = pyg.monthlymean(pyg.standardtimen('1 Jan 2022', 365))`` results in dates that lie on the 1st of the month; the ``units`` are inherited from the underlying time axis, which defaults to `days` in this case.

I do very much think it would be useful to provide some helper functions for creating monthly time axes (e.g. one point per month, which could perhaps be on any day of the month). Monthly mean data is pretty ubiquitous. And that might allow us to handle this situation in the cfmeta decode step in a somewhat more graceful way. For instance, we could use the helper function to create a monthly time axis corresponding to the dates in the file (say assuming the 1st of each month), but set the actual ``units`` attribute to days. But because the behaviour is strictly ambiguous we might want to raise a warning or something?

I don't think we want to start using 'month' or 'year' etc. as actual units for the date calculations.

**Comment by aph42** [comment]
As a follow up; what about the following: we could add helper functions like

`standardmonthlyn(start, n, units='days', ref=None)`

that would return Standard time axis of length `n` all on the same day of the month. So

 `standardmonthlyn('15 Jan 2000', 12)`

would return a 12 element time axis with dates on the 15th of each month in the year 2000.

`standardmonthlyrange(start, end, units = 'days', ref=None, inc=False)`

could do the same for dates spanning from start to end; here I guess we'd have to check whether the day given for start and end matched. Similar helper functions could be added for modeltime365 and modeltime360 as well...

**Comment by admg26** [comment]
I agree with the suggestion for the helper functions. I think we should also make the behaviour of `v = pyg.monthlymean(pyg.standardtimen('1 Jan 2022', 365))` consistent with the helper function. For example, interpotating `v` back to daily data will assume the monthlymean is on 1st of the month while we might want it to assume the 15th. Could we have an option to specify the monthlymean day? For example, `v = pyg.monthlymean(pyg.standardtimen('1 Jan 2022', 365), ref_day=15)`
